### PR TITLE
DL: Add asymmetric cluster check for predict

### PR DESCRIPTION
**Pull request by orhankislal**
JIRA: MADLIB-1394

This commit updates the madlib_keras_predict and associated functions to
use use_gpus parameter like the rest of the keras functions.

Current implementation does not allow for asymmetric gpu configurations.

- [x] Add the module name, JIRA# to PR/commit and description.
- [x] Add tests for the change.

### CODE REVIEW

File: src/ports/postgres/modules/deep_learning/madlib_keras_predict.py_in [hunk 1]
```diff
@@ -244,14 +254,14 @@ class PredictBYOM(BasePredict):
...
+                           images_per_seg, use_gpus, gpus_per_host, segments_per_host,
                            **kwargs):
     SD = kwargs['SD']
     model_key = 'segment_model_predict'
     row_count_key = 'row_count'
     try:
-        device_name = get_device_name_and_set_cuda_env_predict(gpus_per_host,
-                                                       current_seg_id)
+        device_name = get_device_name_and_set_cuda_env(
+            use_gpus, gpus_per_host, current_seg_id)
```
Comment (khannaekta) [inline src/ports/postgres/modules/deep_learning/madlib_keras_predict.py_in]: As we are already validating the gpu configuration prior to calling this function(for fit/evaluate/predict), the function `get_device_name_and_set_cuda_env()` should only care about setting device name (GPU or CPU based on the value of `gpus_per_host`) and should not validate GPU configuration(based on `use_gpus`). 
I think, it's ok to remove `use_gpus` from `get_device_name_and_set_cuda_env()`, unless I am missing some scenario.

File: src/ports/postgres/modules/deep_learning/madlib_keras_predict.py_in [hunk 2]
```diff
@@ -41,22 +41,31 @@ from madlib_keras_wrapper import *
...
-        self.gpus_per_host = gpus_per_host
-        self._set_default_gpus_pred_type()
+        self.module_name = module_name
+
+        self.segments_per_host = get_segments_per_host()
+        self.use_gpus = use_gpus if use_gpus else False
+        if self.use_gpus:
+            accessible_gpus_for_seg = get_accessible_gpus_for_seg(schema_madlib, self.segments_per_host, self.module_name)
+            _assert(len(set(accessible_gpus_for_seg)) == 1,
+                '{0}: Asymetric gpu configurations are not supported'.format(self.module_name))
```
Comment (khannaekta) [inline src/ports/postgres/modules/deep_learning/madlib_keras_predict.py_in]: Asymetric -> Asymmetric

File: src/ports/postgres/modules/deep_learning/madlib_keras_predict.py_in (hunk 1 above)
Comment (orhankislal) [reply src/ports/postgres/modules/deep_learning/madlib_keras_predict.py_in]: I think this is a good suggestion, we can remove use_gpus from this function.

**Comment by fmcquillan99** [comment]
seeing an evaluate error that I don't remember seeing before...
```
DROP TABLE IF EXISTS iris_multi_model, iris_multi_model_summary, iris_multi_model_info;

SELECT madlib.madlib_keras_fit_multiple_model('iris_train_packed',    -- source_table
                                              'iris_multi_model',     -- model_output_table
                                              'mst_table',            -- model_selection_table
                                               6,                     -- num_iterations
                                               FALSE,                 -- use gpus
                                              'iris_test_packed',     -- validation dataset
                                               2,                     -- metrics compute frequency
                                               FALSE,                 -- warm start
                                              'Sophie L.',            -- name
                                              'Simple MLP for iris dataset'  -- description
                                             );


InternalError: (psycopg2.InternalError) plpy.SPIError: plpy.Error: Number of elements in model weights(303) doesn't match model(193). (plpython.c:5038)  (seg0 slice1 10.128.0.41:40000 pid=32019) (plpython.c:5038)
DETAIL:  
Traceback (most recent call last):
  PL/Python function "internal_keras_eval_transition", line 6, in <module>
    return madlib_keras.internal_keras_eval_transition(**globals())
  PL/Python function "internal_keras_eval_transition", line 727, in internal_keras_eval_transition
  PL/Python function "internal_keras_eval_transition", line 125, in set_model_weights
  PL/Python function "internal_keras_eval_transition", line 167, in deserialize_as_nd_weights
  PL/Python function "internal_keras_eval_transition", line 96, in _assert
PL/Python function "internal_keras_eval_transition"
CONTEXT:  Traceback (most recent call last):
  PL/Python function "madlib_keras_fit_multiple_model", line 21, in <module>
    fit_obj = madlib_keras_fit_multiple_model.FitMultipleModel(**globals())
  PL/Python function "madlib_keras_fit_multiple_model", line 42, in wrapper
  PL/Python function "madlib_keras_fit_multiple_model", line 167, in __init__
  PL/Python function "madlib_keras_fit_multiple_model", line 174, in fit_multiple_model
  PL/Python function "madlib_keras_fit_multiple_model", line 200, in train_multiple_model
  PL/Python function "madlib_keras_fit_multiple_model", line 229, in evaluate_model
  PL/Python function "madlib_keras_fit_multiple_model", line 409, in compute_loss_and_metrics
  PL/Python function "madlib_keras_fit_multiple_model", line 691, in get_loss_metric_from_keras_eval
PL/Python function "madlib_keras_fit_multiple_model"
 [SQL: "SELECT madlib.madlib_keras_fit_multiple_model('iris_train_packed',    -- source_table\n                                              'iris_multi_model',     -- model_output_table\n                                              'mst_table',            -- model_selection_table\n                                               6,                     -- num_iterations\n                                               FALSE,                 -- use gpus\n                                              'iris_test_packed',     -- validation dataset\n                                               2,                     -- metrics compute frequency\n                                               FALSE,                 -- warm start\n                                              'Sophie L.',            -- name\n                                              'Simple MLP for iris dataset'  -- description\n                                             );"]
```

**Comment by orhankislal** [comment]
@fmcquillan99 Could you give it another try with the new commit?

**Comment by fmcquillan99** [comment]
@orhankislal yes it does work, thank you for the fix.

I will continue testing now.

**Comment by fmcquillan99** [comment]
this seems too verbose to output:
```
SELECT madlib.load_keras_model('model_arch_library',  -- Output table,
$$
{"class_name": "Sequential", "keras_version": "2.1.6", "config": [{"class_name": "Dense", "config": {"kernel_initializer": {"class_name": "VarianceScaling", "config": {"distribution": "uniform", "scale": 1.0, "seed": null, "mode": "fan_avg"}}, "name": "dense_1", "kernel_constraint": null, "bias_regularizer": null, "bias_constraint": null, "dtype": "float32", "activation": "relu", "trainable": true, "kernel_regularizer": null, "bias_initializer": {"class_name": "Zeros", "config": {}}, "units": 10, "batch_input_shape": [null, 4], "use_bias": true, "activity_regularizer": null}}, {"class_name": "Dense", "config": {"kernel_initializer": {"class_name": "VarianceScaling", "config": {"distribution": "uniform", "scale": 1.0, "seed": null, "mode": "fan_avg"}}, "name": "dense_2", "kernel_constraint": null, "bias_regularizer": null, "bias_constraint": null, "activation": "relu", "trainable": true, "kernel_regularizer": null, "bias_initializer": {"class_name": "Zeros", "config": {}}, "units": 10, "use_bias": true, "activity_regularizer": null}}, {"class_name": "Dense", "config": {"kernel_initializer": {"class_name": "VarianceScaling", "config": {"distribution": "uniform", "scale": 1.0, "seed": null, "mode": "fan_avg"}}, "name": "dense_3", "kernel_constraint": null, "bias_regularizer": null, "bias_constraint": null, "activation": "softmax", "trainable": true, "kernel_regularizer": null, "bias_initializer": {"class_name": "Zeros", "config": {}}, "units": 3, "use_bias": true, "activity_regularizer": null}}], "backend": "tensorflow"}
$$
::json,         -- JSON blob
                               NULL,                  -- Weights
                               'Sophie',              -- Name
                               'MLP with 1 hidden layer'       -- Descr
);
```
produces
```
INFO:  Keras Model Arch: Created new keras model architecture table model_arch_library.
CONTEXT:  PL/Python function "load_keras_model"
INFO:  Keras Model Arch: Added model id 1 to model_arch_library table
CONTEXT:  PL/Python function "load_keras_model"
-[ RECORD 1 ]----+-
load_keras_model | 

Time: 105.684 ms
```
Is this the expected output?

**Comment by orhankislal** [comment]
Yes, those are the expected output and info messages.

**Comment by fmcquillan99** [comment]
It says
`Training set after iteration x`
but does not say
`Validation set after iteration x`
so at first I didn't know that the last 12 were for the validation set.

Is there a way to say `Validation set after iteration x` before listing the data?

```
	Time for training in iteration 4: 9.69201207161 sec
CONTEXT:  PL/Python function "madlib_keras_fit_multiple_model"
INFO:  
	Time for training in iteration 5: 9.86732983589 sec
CONTEXT:  PL/Python function "madlib_keras_fit_multiple_model"
INFO:  
	Time for training in iteration 6: 10.0109279156 sec
DETAIL:  
	Training set after iteration 6:
	mst_key=2: metric=0.958333313465, loss=0.183155477047
	mst_key=11: metric=0.841666638851, loss=0.825849890709
	mst_key=10: metric=0.850000023842, loss=0.270138114691
	mst_key=4: metric=0.891666650772, loss=0.308341085911
	mst_key=3: metric=0.850000023842, loss=0.273101329803
	mst_key=12: metric=0.800000011921, loss=0.981224298477
	mst_key=9: metric=0.816666662693, loss=0.39588457346
	mst_key=1: metric=0.358333319426, loss=1.10315680504
	mst_key=8: metric=0.358333319426, loss=1.09978532791
	mst_key=7: metric=0.691666662693, loss=0.468748778105
	mst_key=5: metric=0.908333361149, loss=0.47638130188
	mst_key=6: metric=0.408333331347, loss=1.00524008274
	mst_key=2: metric=1.0, loss=0.148071616888
	mst_key=11: metric=0.833333313465, loss=0.833429694176
	mst_key=10: metric=0.666666686535, loss=0.476953774691
	mst_key=4: metric=0.733333349228, loss=0.392092883587
	mst_key=3: metric=0.933333337307, loss=0.10978718102
	mst_key=12: metric=0.866666674614, loss=0.987264215946
	mst_key=9: metric=0.600000023842, loss=0.893045544624
	mst_key=1: metric=0.233333334327, loss=1.11105871201
	mst_key=8: metric=0.233333334327, loss=1.14259588718
	mst_key=7: metric=0.566666662693, loss=0.525054514408
	mst_key=5: metric=0.899999976158, loss=0.493363827467
	mst_key=6: metric=0.266666680574, loss=1.04115581512
CONTEXT:  PL/Python function "madlib_keras_fit_multiple_model"
INFO:  
	Time for training in iteration 7: 10.1211631298 sec
CONTEXT:  PL/Python function "madlib_keras_fit_multiple_model"
INFO:  
```

**Comment by fmcquillan99** [comment]
Besides my last comment, LGTM.

**Comment by fmcquillan99** [comment]
looks better
```
	Time for training in iteration 4: 8.73104310036 sec
CONTEXT:  PL/Python function "madlib_keras_fit_multiple_model"
INFO:  
	Time for training in iteration 5: 8.39556002617 sec
CONTEXT:  PL/Python function "madlib_keras_fit_multiple_model"
INFO:  
	Time for training in iteration 6: 8.93426513672 sec
DETAIL:  
	Training set after iteration 6:
	mst_key=2: metric=0.691666662693, loss=0.472569614649
	mst_key=11: metric=0.975000023842, loss=0.136468976736
	mst_key=10: metric=0.949999988079, loss=0.109081745148
	mst_key=4: metric=0.975000023842, loss=0.178229421377
	mst_key=3: metric=0.925000011921, loss=0.184964418411
	mst_key=12: metric=0.975000023842, loss=0.141136452556
	mst_key=9: metric=0.958333313465, loss=0.0976895838976
	mst_key=1: metric=0.375, loss=1.09557545185
	mst_key=8: metric=0.975000023842, loss=0.0692544877529
	mst_key=7: metric=0.958333313465, loss=0.153927713633
	mst_key=5: metric=0.841666638851, loss=0.468217223883
	mst_key=6: metric=0.741666674614, loss=0.504250705242
	Validation set after iteration 6:
	mst_key=2: metric=0.600000023842, loss=0.418071746826
	mst_key=11: metric=1.0, loss=0.0674076601863
	mst_key=10: metric=1.0, loss=0.018841393292
	mst_key=4: metric=1.0, loss=0.101651079953
	mst_key=3: metric=0.933333337307, loss=0.149523466825
	mst_key=12: metric=1.0, loss=0.0722744837403
	mst_key=9: metric=1.0, loss=0.0356690809131
	mst_key=1: metric=0.166666671634, loss=1.12900209427
	mst_key=8: metric=1.0, loss=0.00220014923252
	mst_key=7: metric=1.0, loss=0.0267392881215
	mst_key=5: metric=0.800000011921, loss=0.491473942995
	mst_key=6: metric=0.600000023842, loss=0.556908965111
CONTEXT:  PL/Python function "madlib_keras_fit_multiple_model"
INFO:  
	Time for training in iteration 7: 8.56644105911 sec
CONTEXT:  PL/Python function "madlib_keras_fit_multiple_model"
INFO:  
```

LGTM

**Pull request merged**